client/rest-high-level/src/main/java/org/elasticsearch/client/Request.java

```diff
@@ -307,23 +312,48 @@ static Request update(UpdateRequest updateRequest) throws IOException {
             xContentType = Requests.INDEX_CONTENT_TYPE;
         }
 
-        BytesRef source = XContentHelper.toXContent(updateRequest, xContentType, false).toBytesRef();
-        HttpEntity entity = new ByteArrayEntity(source.bytes, source.offset, source.length, ContentType.create(xContentType.mediaType()));
-
+        HttpEntity entity = createEntity(updateRequest, xContentType);
         return new Request(HttpPost.METHOD_NAME, endpoint, parameters.getParams(), entity);
     }
 
+    static Request search(SearchRequest searchRequest) throws IOException {
+        String endpoint = endpoint(searchRequest.indices(), searchRequest.types(), "_search");
+        Params params = Params.builder();
+        params.putParam(RestSearchAction.TYPED_KEYS_PARAM, "true");
+        params.withRouting(searchRequest.routing());
+        params.withPreference(searchRequest.preference());
+        params.withIndicesOptions(searchRequest.indicesOptions());
+        params.putParam("search_type", searchRequest.searchType().name().toLowerCase(Locale.ROOT));
+        if (searchRequest.requestCache() != null) {
+            params.putParam("request_cache", Boolean.toString(searchRequest.requestCache()));
+        }
+        params.putParam("batched_reduce_size", Integer.toString(searchRequest.getBatchedReduceSize()));
+        if (searchRequest.scroll() != null) {
+            params.putParam("scroll", searchRequest.scroll().keepAlive());
+        }
+        HttpEntity entity = null;
+        if (searchRequest.source() != null) {
+            entity = createEntity(searchRequest.source(), CONTENT_TYPE);
+        }
+        return new Request(HttpGet.METHOD_NAME, endpoint, params.getParams(), entity);
```

Comment:
we should probably introduce some kind of setting in the client that allows to choose whether we should use POST or GET when there's a body.